### PR TITLE
Support encode/decode java.time.Duration and ConfigMemorySize in HOCON

### PR DESCRIPTION
**Pull request by alexmihailov**
- Many use cases require serialization support for `java.time.Duration`. It is not possible to implement a custom serializer in user code, since private methods from `kotlinx.serialization.hocon.Hocon` and `kotlinx.serialization.hocon.AbstractHoconEncoder` are required. To do this, I implemented the `JDurationSerializer` serializer, which can be used in user code.

- Hocon supports the memory size format. [Documentation](https://github.com/lightbend/config/blob/main/HOCON.md#size-in-bytes-format).
In the `com.typesafe.config` library, the `ConfigMemorySize` class is responsible for this functionality. There is no way in user code to implement a serializer for this class. So I've implemented a `ConfigMemorySizeSerializer` that allows to serialize `ConfigMemorySize`.

- The `com.typesafe.config` library can create configuration using Java Bean notation. The `com.typesafe.config.ConfigBeanFactory.create` method is used for this. To support this functionality, I've implemented a `JBeanSerializer` that can only be used to decode configuration into a Java Bean class.

This PR depends on the previous [PR](https://github.com/Kotlin/kotlinx.serialization/pull/2080)

### CODE REVIEW

**Comment by ihostage** [comment]
Hi, Leonid @sandwwraith! 👋  
FYI, this PR very useful to migrate from directly using _Typesafe Config_ to _Kotlinx.Serialization_, because until now _Kotlinx.Serialization_ don't have a few built-in _Typesafe Config_ functionality 🤷‍♂️ 
Thanks, Alex @alexmihaylov! 👍

**Comment by alexmihailov** [comment]
Hi, Leonid @sandwwraith !
I have implemented a public api: `HoconEncoder`, `HoconDecoder`. It really does look a lot better and expands the possibilities for writing custom serializers.
In many cases `java.time.Duration` is used, so I would like to have a serializer for it in the library.
`ConfigMemorySize` is also often used. I also want the serializer for it to be in the library.
Java Bean notation is rarely used and the serializer for it can be removed from the library.

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    // For powers of two.
+    private val memoryUnitFormats = listOf("byte", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB")
+
+    private val valueResolver: (Config, String) -> ConfigMemorySize = { conf, path -> conf.decodeMemorySize(path) }
+
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor("hocon.com.typesafe.config.ConfigMemorySize", PrimitiveKind.STRING)
+
+    override fun deserialize(decoder: Decoder): ConfigMemorySize = if (decoder is HoconDecoder) {
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:42]: Here and in other serializers (and documentation samples for HoconEncoder/Decoder):

IMO, this formatting looks hard to follow because:
1. This is not a one-line `if`, but it starts from the same line as declaration
2. `if` has braces, while `else` does not (even though there's a single statement under `if`).

I suggest writing it from the new line and being consistent with an absence of braces:

```
override fun deserialize(decoder: Decoder): ConfigMemorySize =
    if (decoder is HoconDecoder)
        decoder.decodeValue(valueResolver)
    else
        throw ...
```
(to save some linebreaks, it is possible to collapse `if (decoder is HoconDecoder) decoder.decodeValue(valueResolver)` into one line and `else throw` in another)

If you find braces preferable for some reason, put them both in `if` and `else`.

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * @Serializable
+ * data class Example(
+ *   @Serializable(NestedConfigSerializer::class)
+ *   val d: Config
+ * )
+ *
+ * object NestedConfigSerializer : KSerializer<Config> {
+ *     override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("package.Config", PrimitiveKind.STRING)
+ *
+ *     override fun deserialize(decoder: Decoder): Config = if (decoder is HoconDecoder) {
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt:21]: See my comment for ConfigMemorySizeSerializer

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * val nestedConfig = ConfigFactory.parseString("nested { value = \"test\" }")
+ * val globalConfig = Hocon.encodeToConfig(Example(nestedConfig)) // d: { nested: { value = "test" } }
+ * val newNestedConfig = Hocon.decodeFromConfig(Example.serializer(), globalConfig)
+ * ```
+ */
+@ExperimentalSerializationApi
+sealed interface HoconDecoder {
+
+    /**
+     * Decodes the next value in the current input.
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt:41]: I don't think 'next' is a valid definition here, as Hocon is tree-based (not stream-based like Json). Maybe just say 'Decodes the value at the current path from the input'?

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * Allows calling methods from [Config].
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt:42]:
```suggestion
     * Allows to call methods on a [Config] instance.
```

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+     *
+     * @param E type of serialize value
+     * @param valueResolver lambda for extract serialize value, where conf - original config, path - path expression
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt:45]:
```suggestion
     * @param valueResolver lambda for extracting value, where conf - original config object, path - current path expression being decoded.
```

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * ```
+ */
+@ExperimentalSerializationApi
+sealed interface HoconDecoder {
+
+    /**
+     * Decodes the next value in the current input.
+     * Allows calling methods from [Config].
+     *
+     * @param E type of serialize value
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt:44]:
```suggestion
     * @param E type of value
```

Not sure what you mean here under 'serialize value'

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt [hunk 7]
```diff
@@ -7,17 +7,59 @@ package kotlinx.serialization.hocon
...
-import kotlinx.serialization.builtins.serializer
 import kotlinx.serialization.descriptors.*
 import kotlinx.serialization.encoding.*
+import kotlinx.serialization.hocon.internal.*
 import kotlinx.serialization.internal.*
 import kotlinx.serialization.modules.*
 
+/**
+ * Encoder used by Hocon during serialization.
+ * This interface allows to call methods from the Lightbend/config library on the [Config] object.
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt:18]: This line seems to be copied from the decoder and simply is not true.
WDYT of "This interface allows intercepting serialization process and insertion of arbitrary [ConfigValue] into the output"?

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt [hunk 8]
```diff
@@ -20,3 +20,6 @@ internal fun InvalidKeyKindException(value: ConfigValue) = SerializationExceptio
     "Value of type '${value.valueType()}' can't be used in HOCON as a key in the map. " +
             "It should have either primitive or enum kind."
 )
+
+internal fun UnsupportedFormatException(serializerName: String) =
+    SerializationException("$serializerName must only be applied to Hocon.")
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt:25]:
```suggestion
    SerializationException("$serializerName is supported only in Hocon format.")
```

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt [hunk 9]
```diff
@@ -20,3 +20,6 @@ internal fun InvalidKeyKindException(value: ConfigValue) = SerializationExceptio
     "Value of type '${value.valueType()}' can't be used in HOCON as a key in the map. " +
             "It should have either primitive or enum kind."
 )
+
+internal fun UnsupportedFormatException(serializerName: String) =
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt:24]: You can add `throw` to the function itself for more convenience, just like with e.g. `throwInvalidFloatingPointValueDecoded`: https://github.com/Kotlin/kotlinx.serialization/blob/60c632c970d7da02151fd7c920f40ebc269c0f29/formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonExceptions.kt#L43

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Decode [JDuration] from [Config].
+ * See https://github.com/lightbend/config/blob/main/HOCON.md#duration-format
+ *
+ * @receiver [Config]
+ * @param path in config
+ * @return [JDuration]
+ */
+@SuppressAnimalSniffer
+internal fun Config.decodeDuration(path: String): JDuration = try {
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt:55]: Better to name it as `decodeJavaDuration` to avoid confusion: `encodeDuration` works with Kotlin's one.

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import com.typesafe.config.*
+import java.math.BigInteger
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlinx.serialization.hocon.*
+
+/**
+ * Serializer for [ConfigMemorySize].
+ * For decode using [HOCON Size in bytes format](https://github.com/lightbend/config/blob/main/HOCON.md#size-in-bytes-format).
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:12]: I'm sorry, I can't quite understand what you mean here. Are you saying decoding accepts all possible values from the spec (including powers of two and ten)? And encoding always emits powers of two?

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+ * For encode used format for powers of two: byte, KiB, MiB, GiB, TiB, PiB, EiB, ZiB, YiB.
+ * Encoding use the largest value of format.
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:14]:
```suggestion
 * Encoding use the largest possible integer value.
```

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+import kotlin.time.*
+import kotlinx.serialization.*
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlinx.serialization.hocon.*
+import kotlinx.serialization.hocon.internal.*
+
+/**
+ * Serializer for [JDuration].
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt:14]:
```suggestion
 * Serializer for [java.time.Duration].
```

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+ * For decode using [Duration format][https://github.com/lightbend/config/blob/main/HOCON.md#duration-format].
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt:15]: Same here: hard to understand. You may want to rephrase this as 'All possible Hocon duration formats [link] are accepted for decoding. During encoding, the serializer emits values using time unit short names: ..... The largest integer time unit is encoded.'

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+ * For encode used time unit short names: d, h, m, s, ms, us, ns.
+ * Encoding use the largest time unit.
+ * Example:
+ *      120.seconds -> 2 m;
+ *      121.seconds -> 121 s;
+ *      120.minutes -> 2 h;
+ *      122.minutes -> 122 m;
+ *      24.hours -> 1 d.
+ * When encoding, there is a conversion to [Duration].
+ * All restrictions on the maximum and minimum duration are specified in [Duration].
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt:25]: Do you mean Java duration here? Or is this a copy-paste leftover?

**Review by sandwwraith — CHANGES_REQUESTED** [review]
(i haven't reviewed the tests yet)

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JBeanSerializer.kt [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+ *      age = 27
+ *  }
+ * """.trimIndent())
+ * val exampleBean = Hocon {
+ *  serializersModule = serializersModuleOf(JBeanSerializer.jBeanSerializer<TestJavaBean>())
+ * }.decodeFromConfig(ExampleJavaBean.serializer(), config)
+ * ```
+ */
+@ExperimentalSerializationApi
+class JBeanSerializer<T> @PublishedApi internal constructor(private val clazz: Class<T>): KSerializer<T> {
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JBeanSerializer.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+ *   val duration: java.time.Duration
+ * )
+ * val config = ConfigFactory.parseString("duration = 1 day")
+ * val exampleDuration = Hocon.decodeFromConfig(ExampleDuration.serializer(), config)
+ * val newConfig = Hocon.encodeToConfig(ExampleDuration.serializer(), exampleDuration)
+ * ```
+ */
+@ExperimentalSerializationApi
+@SuppressAnimalSniffer
+object JDurationSerializer : KSerializer<JDuration> {
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+     * @param valueResolver lambda for extract serialize value, where conf - original config, path - path expression
+     * @return serialize value [E]
+     */
+    fun <E> decodeValue(valueResolver: (conf: Config, path: String) -> E): E
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 19]
```diff
@@ -0,0 +1,49 @@
+package kotlinx.serialization.hocon
+
+import com.typesafe.config.Config
+import kotlinx.serialization.ExperimentalSerializationApi
+
+/**
+ * Decoder used by Hocon during deserialization.
+ * This interface allows to call methods from the Lightbend/config library on the [Config] object.
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt (hunk 1 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt (hunk 2 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt (hunk 3 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt]: Agree with you

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt (hunk 4 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt (hunk 5 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt (hunk 6 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt (hunk 7 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt]: Yes, that's what I meant

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt (hunk 8 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt (hunk 9 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconExceptions.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt (hunk 10 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt (hunk 11 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:12]: Yes, when decoding, can use all formats from the specification. For encoding, I selected on one option - to use powers of two, as it is easier to implement.

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt (hunk 12 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt (hunk 13 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt (hunk 14 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/JDurationSerializer.kt]: Fixed

File: formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconMemorySizeTest.kt [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+        val ln: List<Nullable>,
+        val f: Boolean,
+        val ld: List<ConfigMemorySize>,
+        val mp: Map<String, ConfigMemorySize>,
+        val mpp: Map<ConfigMemorySize, ConfigMemorySize>
+    )
+
+    @Test
+    fun testSerializeMemorySize() {
+        Hocon.encodeToConfig(Simple(ofBytes(10))).assertContains("size = 10 byte")
```
Comment (sandwwraith) [inline formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconMemorySizeTest.kt:48]: I think it makes sense to test also deserialization here via some helper function to avoid repeating `encodeToConfig`, like

```
val res = Hocon.encodeToConfig(simple)
res.assertContains(s)
assertEquals(simple, Hocon.decodeFromConfig(res))
```

**Review by sandwwraith — CHANGES_REQUESTED** [review]

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+@ExperimentalSerializationApi
+sealed interface HoconDecoder {
+
+    /**
+     * Decodes the value at the current path from the input.
+     * Allows to call methods on a [Config] instance.
+     *
+     * @param E type of value
+     * @param extractValueAtPath lambda for extracting value, where conf - original config object, path - current path expression being decoded.
+     * @return serialize value [E]
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconDecoder.kt:44]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt [hunk 22]
```diff
@@ -7,17 +7,59 @@ package kotlinx.serialization.hocon
...
+ *
+ * object NestedConfigSerializer : KSerializer<Config> {
+ *     override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("package.Config", PrimitiveKind.STRING)
+ *
+ *     override fun deserialize(decoder: Decoder): Config = if (decoder is HoconDecoder) {
+ *         decoder.decodeValue { conf, path -> conf.getConfig(path) // call to Lightbend/config library  }
+ *     } else throw SerializationException("This class can be decoded only by Hocon format")
+ *
+ *     override fun serialize(encoder: Encoder, value: Config) {
+ *         if (encoder is AbstractHoconEncoder) {
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt:36]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt [hunk 23]
```diff
@@ -7,17 +7,59 @@ package kotlinx.serialization.hocon
...
+ * val newNestedConfig = Hocon.decodeFromConfig(Example.serializer(), globalConfig)
+ * ```
+ */
+@ExperimentalSerializationApi
+sealed interface HoconEncoder {
+
+    /**
+     * Appends the given [ConfigValue] element to the current output.
+     *
+     * @param value [ConfigValue]
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt:53]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt [hunk 24]
```diff
@@ -7,17 +7,59 @@ package kotlinx.serialization.hocon
...
+ *     }
+ * }
+ *
+ * val nestedConfig = ConfigFactory.parseString("nested { value = \"test\" }")
+ * val globalConfig = Hocon.encodeToConfig(Example(nestedConfig)) // d: { nested: { value = "test" } }
+ * val newNestedConfig = Hocon.decodeFromConfig(Example.serializer(), globalConfig)
+ * ```
+ */
+@ExperimentalSerializationApi
+sealed interface HoconEncoder {
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt:48]: The `HoconEncoder.kt` file already existed when the `HoconEncoder` interface was created. I can put the `HoconEncoder` interface into a separate file, and put the rest of the classes into a `HoconEncoders.kt` or `Encoders.kt` file. Leonid, how do you think ?

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Encode [Duration] objects using time unit short names: d, h, m, s, ms, us, ns.
+ * Example:
+ *     120.seconds -> 2 m;
+ *     121.seconds -> 121 s;
+ *     120.minutes -> 2 h;
+ *     122.minutes -> 122 m;
+ *     24.hours -> 1 d.
+ * Encoding use the largest time unit.
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt:18]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+import java.math.BigInteger
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlinx.serialization.hocon.*
+
+/**
+ * Serializer for [ConfigMemorySize].
+ * For decode using [HOCON Size in bytes format](https://github.com/lightbend/config/blob/main/HOCON.md#size-in-bytes-format).
+ * For encode used format for powers of two: byte, KiB, MiB, GiB, TiB, PiB, EiB, ZiB, YiB.
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:13]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+            var bytes = value.toBytesBigInteger()
+            var unitIndex = 0
+            while (bytes.and(andVal) == BigInteger.ZERO) { // n & 0x3ff == 0
+                if (unitIndex < memoryUnitFormats.lastIndex) {
+                    bytes = bytes.shiftRight(10)
+                    unitIndex++
+                } else break
+            }
+            encoder.encodeString("$bytes ${memoryUnitFormats[unitIndex]}")
+        } else throwUnsupportedFormatException("ConfigMemorySizeSerializer")
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:60]: Fixed

File: formats/hocon/src/test/java/kotlinx/serialization/hocon/TestJavaBean.java [hunk 28]
```diff
@@ -0,0 +1,39 @@
+package kotlinx.serialization.hocon;
+
+import java.util.Objects;
+
+public class TestJavaBean {
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/test/java/kotlinx/serialization/hocon/TestJavaBean.java:5]: Yes, removed this test class

File: formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconJavaDurationTest.kt [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+    @Test
+    fun testSerializeComplexDuration() {
+        val obj = Complex(
+            i = 6,
+            s = Simple(ofMinutes(5)),
+            n = Nullable(null),
+            l = listOf(Simple(ofMinutes(1)), Simple(ofSeconds(2))),
+            ln = listOf(Nullable(null), Nullable(ofHours(6))),
+            f = true,
+            ld = listOf(ofDays(1), ofMinutes(1), ofNanos(5)),
```
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconJavaDurationTest.kt:99]: I removed the `@Contextual` annotation.

File: formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconMemorySizeTest.kt (hunk 20 above)
Comment (alexmihailov) [reply formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconMemorySizeTest.kt:48]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+ * All restrictions on the maximum and minimum duration are specified in [Duration].
+ * @param value [Duration]
```
[the remark this answers is not in the document]
Comment (sandwwraith) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt:20]: Bump

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt [hunk 31]
```diff
@@ -0,0 +1,65 @@
...
+    }
+}
+
+/**
+ * Decode [JDuration] from [Config].
+ * See https://github.com/lightbend/config/blob/main/HOCON.md#duration-format
+ *
+ * @receiver [Config]
+ * @param path in config
+ * @return [JDuration]
```
[the remark this answers is not in the document]
Comment (sandwwraith) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/internal/HoconDuration.kt:52]: Bump

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt (hunk 11 above)
Comment (sandwwraith) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:12]: Bump: Rewrite as 'All possible Hocon size formats....'

File: formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconJavaDurationTest.kt [hunk 32]
```diff
@@ -0,0 +1,188 @@
...
+        val ln: List<Nullable>,
+        val f: Boolean,
+        val ld: List<Duration>,
+        val mp: Map<String, Duration>,
+        val mpp: Map<Duration, Duration>
+    )
+
+    @Test
+    fun testSerializeDuration() {
+        Hocon.encodeToConfig(Simple(ofMinutes(10))).assertContains("d = 10 m")
```
Comment (sandwwraith) [inline formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconJavaDurationTest.kt:44]: Same here: why don't write function similar to `testMemorySize`?

**Comment by sandwwraith** [comment]
@alexmihailov Hi, do you have time to finish this on this week? I plan to release 1.5.0 soon.

**Comment by alexmihailov** [comment]
Hello! Thursday/Friday I will try to finish

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt (hunk 11 above)
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:12]: Fixed

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt (hunk 24 above)
[the remark this answers is not in the document]
Comment (alexmihailov) [reply formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/HoconEncoder.kt:45]: I moved the HoconEncoder interface to a separate file

File: formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconJavaDurationTest.kt (hunk 32 above)
Comment (alexmihailov) [reply formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconJavaDurationTest.kt:44]: Implemented the testJavaDuration function

**Comment by alexmihailov** [comment]
@sandwwraith Hi! I checked the hidden sections of conversations and made corrections. I'm sorry I didn't notice right away.

File: formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlinx.serialization.hocon.*
+
+/**
+ * Serializer for [ConfigMemorySize].
+ * All possible Hocon size formats [https://github.com/lightbend/config/blob/main/HOCON.md#size-in-bytes-format] are accepted for decoding.
+ * During encoding, the serializer emits values using powers of two: byte, KiB, MiB, GiB, TiB, PiB, EiB, ZiB, YiB.
+ * Encoding use the largest possible integer value.
```
Comment (sandwwraith) [inline formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/serializers/ConfigMemorySizeSerializer.kt:14]:
```suggestion
 * Encoding uses the largest possible integer value.
```

**Review by sandwwraith — APPROVED** [review]
Thank you very much for all your time!
